finite_approx/gibbs_sampler_lib.py

```diff
@@ -101,3 +103,214 @@ def sample(self, burnin, num_gibbs_draws):
+                p_z_nk1_cond_Z = (np.sum(self.z[:, k]) - self.z[n, k] + \
+                                self.alpha/self.k_approx)\
+                                /(self.x_n + self.alpha/self.k_approx)
+                assert (p_z_nk1_cond_Z >= 0) & (p_z_nk1_cond_Z <= 1)
+
+                if self.z[n,k] == 1:
+                    p_znk_cond_z_stay = p_z_nk1_cond_Z
+                    p_znk_cond_z_update = 1 - p_z_nk1_cond_Z
+                else:
+                    p_znk_cond_z_update = p_z_nk1_cond_Z
+                    p_znk_cond_z_stay = 1 - p_z_nk1_cond_Z
+
+                # next compute p(X|Z_new)
+                # equation (8) in Griffiths and Ghahramani
+                [log_lh_X_update, var_inv_update] = update_x_lp_cond_z(\
+                    self.x, self.z, var_inv, self.sigma_eps, self.sigma_a, \
+                    self.k_approx, n, k)
+
+
+                log_p_zn_update = log_lh_X_update + np.log(p_znk_cond_z_update) \
+                    - sp.misc.logsumexp([log_lh_X_update + np.log(p_znk_cond_z_update),\
+                        log_lh_X_stay+ np.log(p_znk_cond_z_stay)])
+
+                assert (np.exp(log_p_zn_update) >= 0) & (np.exp(log_p_zn_update) <= 1)
+
+                choice = np.random.binomial(1, np.exp(log_p_zn_update))
+
+                # update Z and prep log likelihood for next round:
+                if choice == 1: # flip Z_nk
+                    log_lh_X_stay = deepcopy(log_lh_X_update)
+                    var_inv = deepcopy(var_inv_update)
+                    self.z[n,k] = 1 - self.z[n,k]
+                else:
+                    #log_lh_X_stay = deepcopy(log_lh_X_stay)
+                    #var_inv = deepcopy(var_inv)
+                    #self.z[n,k] = self.z[n,k]
+                    pass
+
+        if keep_draw:
+            self.z_draws.append(self.z)
+
+    def sample(self, burnin, num_gibbs_draws):
+
+        print('Sampling:')
+        bar = progressbar.ProgressBar(max_value=num_gibbs_draws + burnin)
+        for n in bar(range(num_gibbs_draws + burnin)):
+            self.draw_z(keep_draw = n >= burnin)
+
+        print('Done sampling :)')
+
+
+def update_x_lp_cond_z(X, Z, Var_inv, sigma_eps, sigma_A, K_approx, n,k):
+# likelihood p(X|Z)-- equation (8) in Griffiths and Ghahramani
+# http://mlg.eng.cam.ac.uk/zoubin/papers/ibp-nips05.pdf
+# outputs the likelihood at Z  when Z has component n,k flipped
+# Var_inv refers to inv(Z^T * Z + sigma_eps/sigma_A I), the previous inverse
+
+    D = np.shape(X)[1]
+    N = np.shape(X)[0]
+    K = np.shape(Z)[1]
+
+    assert np.shape(X)[0] == np.shape(Z)[0]
+
+    Z_update = deepcopy(Z)
+    Z_update[n,k] = 1 - Z[n,k]
+
+    inv_var_flip = update_inv_var(Z, Var_inv, sigma_eps, sigma_A, n, k)
+
+    mean_A = np.dot(np.dot(inv_var_flip, Z_update.T), X)
+
+    log_likelihood = -1/(2*sigma_eps) * \
+            np.trace(np.dot(X.T, X - np.dot(Z_update, mean_A)) )
+
+    const = np.linalg.det(inv_var_flip)**(-D/2)
```

Comment:
Same as above, for numerical stability calculate the log determinant directly rather than the determinant and take the log.